source/common/network/lc_trie.h

```diff
@@ -146,27 +158,108 @@ class LcTrie {
+    uint32_t length_{0};
+    // Tag(s) for this entry.
+    TagSet tags_;
+  };
+
+  /**
+   * Binary trie used to simplify the construction of Level Compressed Tries.
+   * This data type supports two operations:
+   *   1. Add a prefix to the trie.
+   *   2. Push the prefixes the leaves of the trie.
+   * That second operation produces a new set of prefixes that yield the same
+   * match results as the original set of prefixes from which the BinaryTrie
+   * was constructed, but with an important difference: the new prefixes are
+   * guaranteed not to be nested within each other. That allows the use of the
+   * classic LC Trie construction algorithm, which is fast and (relatively)
+   * simple but does not work properly with nested prefixes.
+   */
+  template <class IpType, uint32_t address_size = CHAR_BIT * sizeof(IpType)> class BinaryTrie {
+  public:
+    BinaryTrie() : root_(std::make_unique<Node>()) {}
+
+    /**
+     * Add a CIDR prefix and associated tag to the binary trie. If an entry already
+     * exists for the prefix, merge the tag into the existing entry.
+     */
+    void insert(const IpPrefix<IpType>& prefix) {
+      Node* node = root_.get();
+      for (uint32_t i = 0; i < prefix.length_; i++) {
+        auto bit = static_cast<uint32_t>(extractBits(i, 1, prefix.ip_));
+        NodePtr& next_node = node->children[bit];
+        if (next_node == nullptr) {
+          next_node = std::make_unique<Node>();
+        }
+        node = next_node.get();
+      }
+      if (node->tags == nullptr) {
+        node->tags = std::make_shared<TagSet>();
+      }
+      node->tags->insert(prefix.tags_.begin(), prefix.tags_.end());
+    }
+
+    /**
+     * Update each node in the trie to inherit/override its ancestors' tags,
+     * and then push the prefixes in the binary trie to the leaves so that:
+     *  1) each leaf contains a prefix, and
+     *  2) no leaf's prefix is nested within another leaf's prefix.
+     * @return the prefixes associated with the leaf nodes.
+     */
+    std::vector<IpPrefix<IpType>> push_leaves() {
+      std::vector<IpPrefix<IpType>> prefixes;
+      std::function<void(Node*, TagSetSharedPtr, unsigned, IpType)> visit =
+          [&](Node* node, TagSetSharedPtr tags, unsigned depth, IpType prefix) {
+            // Inherit any tags set by ancestor nodes.
+            if (tags != nullptr) {
+              if (node->tags == nullptr) {
+                node->tags = tags;
+              } else {
+                node->tags->insert(tags->begin(), tags->end());
+              }
+            }
+            // If a node has exactly one child, create a second child node
+            // that inherits the union of all tags set by any ancestor nodes.
+            // This gives the trie an important new property: all the configured
+            // prefixes end up at the leaves of the trie. As no leaf is nested
+            // under another leaf (or one of them would not be a leaf!), the
```

Comment:
This seems related to the above comment. I think I get what you're saying here; maybe just think about rewording slightly the above comment to make it clearer with an example.

Reply:
Can you elaborate a bit on what type of example you're looking for?